tls/s2n_cipher_suites.c

```diff
@@ -1168,27 +1171,57 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
     return 0;
 }
 
+bool s2n_cipher_suite_uses_chacha20_alg(struct s2n_cipher_suite *cipher_suite) {
+    return cipher_suite && cipher_suite->record_alg && cipher_suite->record_alg->cipher == &s2n_chacha20_poly1305;
+}
+
+static S2N_RESULT s2n_client_signalled_chacha20_boosting(struct s2n_stuffer wire_stuffer, uint32_t cipher_suite_len, bool *client_signalled_chacha20_boosting) {
+    RESULT_ENSURE_REF(client_signalled_chacha20_boosting);
+    *client_signalled_chacha20_boosting = false;
+
+    uint8_t clients_first_cipher_iana[S2N_SSLv2_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&wire_stuffer, clients_first_cipher_iana, cipher_suite_len));
+
+    struct s2n_cipher_suite *client_first_cipher_suite = NULL;
+    RESULT_GUARD(s2n_cipher_suite_from_iana(clients_first_cipher_iana, &client_first_cipher_suite));
+    RESULT_ENSURE_REF(client_first_cipher_suite);
+
+    if (s2n_cipher_suite_uses_chacha20_alg(client_first_cipher_suite)) {
+        *client_signalled_chacha20_boosting = true;
+    }
```

Comment:
s2n_cipher_suite_from_iana is pretty expensive, and in this case we only care about chacha20 ciphers, not all ciphers. Is there any way to make this check more efficient, without re-introducing the potential for a list of chacha20 ciphers to get out of date?

This isn't a hard blocker; this does work and we do already do this conversion at runtime for session tickets.

Reply:
I'll try looking into optimising this in a separate PR.

Reply:
To be clear: I'm not asking you to optimize s2n_cipher_suite_from_iana. There are simpler, more targeted solutions.